File: CS/RichEditPopupMenu/Form1.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Windows.Forms;
#region #usings
using DevExpress.Utils.Menu;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Menu;
using DevExpress.XtraSpellChecker;
#endregion #usings

namespace RichEditPopupMenuExample {
    public partial class Form1 : Form {
        static List<char> wordSeparators = CreateWordSeparators();

        static List<char> CreateWordSeparators() {
            List<char> result = new List<char>();
            result.Add(' ');
            result.Add('\t');
            result.Add('\n');
            result.Add('\r');
            result.Add(',');
            result.Add('.');
            result.Add('-');
            result.Add('(');
            result.Add(')');
            result.Add('{');
            result.Add('}');
            result.Add('[');
            result.Add(']');
            result.Add('"');
            result.Add('\'');
            result.Add('<');
            result.Add('>');
            result.Add(':');
            result.Add(';');
            result.Add('\\');
            result.Add('/');
            return result;
        }

        public Form1() {
            InitializeComponent();
            richEditControl1.PopupMenuShowing+=new DevExpress.XtraRichEdit.PopupMenuShowingEventHandler(richEditControl1_PopupMenuShowing);
        }

#region #popupmenushowing
private void richEditControl1_PopupMenuShowing(object sender, 
    DevExpress.XtraRichEdit.PopupMenuShowingEventArgs e)
{
    if (spellChecker1.SpellCheckMode == SpellCheckMode.OnDemand)
    {
        DocumentPosition pos = richEditControl1.Document.CaretPosition;
        int wordEnd = GetWordEndIndex(pos);
        int wordStart = GetWordStartIndex(pos);
        if (wordEnd <= wordStart)
            return;
        DocumentRange range = richEditControl1.Document.CreateRange(wordStart, wordEnd - wordStart);
        string word = richEditControl1.Document.GetText(range);
        if (spellChecker1.IsMisspelledWord(word, spellChecker1.Culture))
            CreateMenuItems(e.Menu, range, word);
    }
}
void CreateMenuItems(RichEditPopupMenu menu, DocumentRange range, string word) {
    SuggestionCollection suggestions = this.spellChecker1.GetSuggestions(word);
    int count = suggestions.Count;
    if (count > 0) {
        int lastIndex = Math.Min(count - 1, 5);
        for (int i = lastIndex; i >= 0; i--) {
            SuggestionBase suggestion = suggestions[i];
            SuggestionMenuItem item = 
                new SuggestionMenuItem(this.richEditControl1.Document, suggestion.Suggestion, range);
            item.Image = RichEditPopupMenuExample.Properties.Resources.suggestion;
            menu.Items.Insert(0, item);
        }
    }
    else {
        DXMenuItem emptyItem = new DXMenuItem("no spelling suggestions");
        emptyItem.Enabled = false;
        menu.Items.Insert(0, emptyItem);
    }
}
#endregion #popupmenushowing

        char GetCharacter(int position) {
            DocumentRange range = this.richEditControl1.Document.CreateRange(position, 1);
            return this.richEditControl1.Document.GetText(range)[0];
        }
        int GetWordEndIndex(DocumentPosition position) {
            int currentPosition = position.ToInt();
            //int result = currentPosition;
            int endPosition = this.richEditControl1.Document.Range.End.ToInt()-1;
            while (currentPosition <= endPosition && !wordSeparators.Contains(GetCharacter(currentPosition))) {
                currentPosition++;
                //result = currentPosition;
            }
            return currentPosition;
        }
        int GetWordStartIndex(DocumentPosition position) {
            int currentPosition = position.ToInt();
            int result = currentPosition;
            while (currentPosition >= 0 && !wordSeparators.Contains(GetCharacter(currentPosition))) {
                result = currentPosition;
                currentPosition--;
            }
            return result;
        }
        void Form1_Load(object sender, EventArgs e) {
            CultureInfo usCulture = new CultureInfo("en-US");
            SpellCheckerISpellDictionary dictionary1 = new SpellCheckerISpellDictionary(@"..\..\american.xlg", @"..\..\english.aff", usCulture);
            dictionary1.AlphabetPath = @"..\..\EnglishAlphabet.txt";
            dictionary1.Load();
            this.spellChecker1.Dictionaries.Add(dictionary1);

            this.richEditControl1.Text = @"Accordnig to an englnsih unviersitry sutdy the oredr of letetrs in a word dosen't mttaer, the olny thnig thta's imporantt is that the frsit and last ltteer of eevry word is in the crrecot psoition. The rset can be jmbueld and one is stlil able to read the txet withuot dificultfiy.";
            this.labelControl1.Text = "Right-click a misspelled word for a context menu containing a list of suggested replacements"; 
        }

    }
#region #menuitem
public class SuggestionMenuItem : DXMenuItem {
    readonly Document document;
    readonly DocumentRange range;

    public SuggestionMenuItem(Document document, string suggestion, DocumentRange range)
        : base(suggestion) {
        this.document = document;
        this.range = range;
    }

    protected override void OnClick() {
        document.Replace(range, Caption);
    }
}
#endregion #menuitem
}

File: VB/RichEditPopupMenu/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Windows.Forms
#Region "#usings"
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Menu
Imports DevExpress.XtraSpellChecker
#End Region ' #usings

Namespace RichEditPopupMenuExample
	Partial Public Class Form1
		Inherits Form
		Private Shared wordSeparators As List(Of Char) = CreateWordSeparators()

		Private Shared Function CreateWordSeparators() As List(Of Char)
			Dim result As New List(Of Char)()
			result.Add(" "c)
			result.Add(ControlChars.Tab)
			result.Add(ControlChars.Lf)
			result.Add(ControlChars.Cr)
			result.Add(","c)
			result.Add("."c)
			result.Add("-"c)
			result.Add("("c)
			result.Add(")"c)
			result.Add("{"c)
			result.Add("}"c)
			result.Add("["c)
			result.Add("]"c)
			result.Add(""""c)
			result.Add("'"c)
			result.Add("<"c)
			result.Add(">"c)
			result.Add(":"c)
			result.Add(";"c)
			result.Add("\"c)
			result.Add("/"c)
			Return result
		End Function

		Public Sub New()
			InitializeComponent()
			AddHandler richEditControl1.PopupMenuShowing, AddressOf richEditControl1_PopupMenuShowing
		End Sub

#Region "#popupmenushowing"
Private Sub richEditControl1_PopupMenuShowing(ByVal sender As Object,  _
ByVal e As DevExpress.XtraRichEdit.PopupMenuShowingEventArgs)
	If spellChecker1.SpellCheckMode = SpellCheckMode.OnDemand Then
		Dim pos As DocumentPosition = richEditControl1.Document.CaretPosition
		Dim wordEnd As Integer = GetWordEndIndex(pos)
		Dim wordStart As Integer = GetWordStartIndex(pos)
		If wordEnd <= wordStart Then
			Return
		End If
		Dim range As DocumentRange =  _
richEditControl1.Document.CreateRange(wordStart, wordEnd - wordStart)
		Dim word As String = richEditControl1.Document.GetText(range)
		If spellChecker1.IsMisspelledWord(word, spellChecker1.Culture) Then
			CreateMenuItems(e.Menu, range, word)
		End If
	End If
End Sub
Private Sub CreateMenuItems(ByVal menu As RichEditPopupMenu, ByVal range As DocumentRange, ByVal word As String)
	Dim suggestions As SuggestionCollection = Me.spellChecker1.GetSuggestions(word)
	Dim count As Integer = suggestions.Count
	If count > 0 Then
		Dim lastIndex As Integer = Math.Min(count - 1, 5)
		For i As Integer = lastIndex To 0 Step -1
		   Dim suggestion As SuggestionBase = suggestions(i)
		   Dim item As New SuggestionMenuItem(Me.richEditControl1.Document, suggestion.Suggestion, range)
		   item.Image = Resources.suggestion
		   menu.Items.Insert(0, item)
		Next i
	Else
		Dim emptyItem As New DXMenuItem("no spelling suggestions")
		emptyItem.Enabled = False
		menu.Items.Insert(0, emptyItem)
	End If
End Sub
#End Region ' #popupmenushowing

		Private Function GetCharacter(ByVal position As Integer) As Char
			Dim range As DocumentRange = Me.richEditControl1.Document.CreateRange(position, 1)
			Return Me.richEditControl1.Document.GetText(range)(0)
		End Function
		Private Function GetWordEndIndex(ByVal position As DocumentPosition) As Integer
			Dim currentPosition As Integer = position.ToInt()
			'int result = currentPosition;
			Dim endPosition As Integer = Me.richEditControl1.Document.Range.End.ToInt()-1
			Do While currentPosition <= endPosition _
 AndAlso Not wordSeparators.Contains(GetCharacter(currentPosition))
				currentPosition += 1
				'result = currentPosition;
			Loop
			Return currentPosition
		End Function
		Private Function GetWordStartIndex(ByVal position As DocumentPosition) As Integer
			Dim currentPosition As Integer = position.ToInt()
			Dim result As Integer = currentPosition
			Do While currentPosition >= 0 AndAlso Not wordSeparators.Contains(GetCharacter(currentPosition))
				result = currentPosition
				currentPosition -= 1
			Loop
			Return result
		End Function
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Dim usCulture As New CultureInfo("en-US")
			Dim dictionary1 As New SpellCheckerISpellDictionary("..\..\american.xlg", "..\..\english.aff", usCulture)
			dictionary1.AlphabetPath = "..\..\EnglishAlphabet.txt"
			dictionary1.Load()
			Me.spellChecker1.Dictionaries.Add(dictionary1)

			Me.richEditControl1.Text = "Accordnig to an englnsih unviersitry sutdy the oredr of letetrs in a word dosen't mttaer, the olny thnig thta's imporantt is that the frsit and last ltteer of eevry word is in the crrecot psoition. The rset can be jmbueld and one is stlil able to read the txet withuot dificultfiy."
			Me.labelControl1.Text = "Right-click a misspelled word for a context menu containing a list of suggested replacements"
		End Sub

	End Class
#Region "#menuitem"
Public Class SuggestionMenuItem
	Inherits DXMenuItem
	Private ReadOnly document As Document
	Private ReadOnly range As DocumentRange

	Public Sub New(ByVal document As Document, ByVal suggestion As String, ByVal range As DocumentRange)
		MyBase.New(suggestion)
		Me.document = document
		Me.range = range
	End Sub

	Protected Overrides Sub OnClick()
		document.Replace(range, Caption)
	End Sub
End Class
#End Region ' #menuitem
End Namespace
